docs/RFCS/online_schema_change.md

```diff
@@ -0,0 +1,144 @@
+is not used for read operations. Lastly, the `PUBLIC` state allows the
+index or column to be used normally.
+
+The state for an index or column moves from `DELETE_ONLY` to
+`WRITE_ONLY` to `PUBLIC` or in reverse. It is invalid to jump from the
+`DELETE_ONLY` to `PUBLIC` state or vice versa. By moving the state in
+these discrete steps it can be shown that that a cluster will not
+create invalid data when two consecutive versions of the descriptor
+are in use concurrently. The table lease mechanism will be used to
+ensure that there are only two active versions of a descriptor in use
+in a cluster and the scheme change implementation will ensure that
+these two versions follow the state transition invariant.
+
+When an index is added to a table a potentially time consuming
+backfill operation needs to be performed to create the index
+data. This operation is performed when the index is in the
+`WRITE_ONLY` state and completed before moving the index to the
+`PUBLIC` state. Backfilling creates index entries for every row in the
+table. As such, it is infeasible for this backfilling to be performed
+transactionally: the transaction would create an unreasonable number
+of write intents and user transactions could cause the transaction to
+abort. Instead, backfill be performed as a series of small
+transactions that process a fraction of the table at a time:
+
+```go
+  startKey := MakeTableIndexKey(desc.ID, desc.Primary)
+  endKey := startKey.PrefixEnd()
+  for startKey != endKey {
+    var lastKey roachpb.Key
+    err := db.Txn(func(txn *Txn) error {
+      txn.SetPriority(VeryLowPriority)
+      scan, err := txn.Scan(startKey, endKey, 1000)
+      if err != nil {
+        return err
+      }
+      lastKey = getLastKeyOfFullRowInScan(scan)
+      b := makeIndexKeysBatch(scan)
+      return txn.CommitInBatch(b)
+    })
+    if err != nil {
+      // Abort!
+    }
+    startKey = lastKey.Next()
+  }
+```
+
+The above pseudo-code is intended to give the gist of how backfilling
+will work. We'll iterate over the primary indexn transactionally
+retrieving keys and generating index entries. We perform the
+backfilling work transactionally in order to avoid anomalies involving
+deletions. Backfilling might duplicate work performed by concurrent
+insertions or updates, but the blind overwrites of identical data are
+safe.
+
+Since schema change operations are potentially long running they need
+to be restartable or abortable if the node performing them dies. We
+accomplish this by performing the schema change operation for a table
+on a well known node: the replica holding the leader lease for the
+first range of the table (i.e. containing the key `/<tableID>`). When
+a node receives a schema change operation such as `CREATE INDEX` it
+will forward the operation to this "table leader". When the table
+leader restarts it will load the associated table descriptor and
+restart or abort the schema change operation.
```

Comment:
Yes, at least my impression was that you never just get to 'delete' a newer
descriptor but instead you add a new one on top. If that's not the plan, an
outline for aborting would be appreciated.
On Oct 15, 2015 14:29, "Peter Mattis" <notifications@github.com> wrote:

> In docs/RFCS/online_schema_change.md
> <https://github.com/cockroachdb/cockroach/pull/2842#discussion_r42117376>:
>
> > +will work. We'll iterate over the primary indexn transactionally
> > +retrieving keys and generating index entries. We perform the
> > +backfilling work transactionally in order to avoid anomalies involving
> > +deletions. Backfilling might duplicate work performed by concurrent
> > +insertions or updates, but the blind overwrites of identical data are
> > +safe.
> > +
> > +Since schema change operations are potentially long running they need
> > +to be restartable or abortable if the node performing them dies. We
> > +accomplish this by performing the schema change operation for a table
> > +on a well known node: the replica holding the leader lease for the
> > +first range of the table (i.e. containing the key `/<tableID>`). When
> > +a node receives a schema change operation such as `CREATE INDEX` it
> > +will forward the operation to this "table leader". When the table
> > +leader restarts it will load the associated table descriptor and
> > +restart or abort the schema change operation.
>
> By "forward schema change" do you mean it will cause the descriptor
> version to increase?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/2842/files#r42117376>.
>